python/fast_mlsirm/multilevel/estimation.py

```
@@ -127,4 +197,400 @@ def weighted_contextual_effect(
+        for occasion in grouped[respondent_id]:
+            time_offsets.append(occasion.time_offset_milliseconds)
+            ordered_occasions.append(occasion)
+        row_offsets.append(len(time_offsets))
+    return row_offsets, ordered_occasions, time_offsets
+
+
+def _observed_value(values: Mapping[str, float], occasion_id: str) -> float:
+    """Return one caller observation as a real float, or NaN when absent."""
+    try:
+        raw = values.get(occasion_id, np.nan)
+    except Exception:
```

Ruff (0.16.1)

[warning] Do not catch blind exception: `Exception`

(BLE001)